src/codegen.cpp

```diff
@@ -4458,6 +4472,51 @@ static jl_cgval_t emit_call(jl_codectx_t &ctx, jl_expr_t *ex, jl_value_t *rt, bo
         }
     }
 
+    // handle calling an OpaqueClosure
+    if (jl_is_concrete_type(f.typ) && jl_subtype(f.typ, (jl_value_t*)jl_opaque_closure_type)) {
+        jl_value_t *oc_argt = jl_tparam0(f.typ);
+        jl_value_t *oc_rett = jl_tparam1(f.typ);
+        if (jl_tupletype_length_compat(oc_argt, nargs-1)) {
+            jl_svec_t *types = jl_get_fieldtypes((jl_datatype_t*)oc_argt);
+            size_t ntypes = jl_svec_len(types);
+            for (size_t i = 0; i < nargs-1; ++i) {
+                jl_value_t *typ = i >= ntypes ? jl_svecref(types, ntypes-1) : jl_svecref(types, i);
+                if (jl_is_vararg(typ))
+                    typ = jl_unwrap_vararg(typ);
+                emit_typecheck(ctx, argv[i+1], typ, "typeassert");
+                argv[i+1] = update_julia_type(ctx, argv[i+1], typ);
+            }
+            jl_returninfo_t::CallingConv cc = jl_returninfo_t::CallingConv::Boxed;
+            unsigned return_roots = 0;
+
+            // Load specptr
+            Type *pint8 = getInt8PtrTy(ctx.builder.getContext());
+            jl_cgval_t &theArg = argv[0];
+            Value *argaddr = NULL;
+            if (theArg.ispointer()) {
+                argaddr = emit_bitcast(ctx, data_pointer(ctx, theArg), pint8);
+            } else {
+                Type *to = julia_type_to_llvm(ctx, f.typ);
+                argaddr = emit_static_alloca(ctx, to);
+                jl_aliasinfo_t ai = jl_aliasinfo_t::fromTBAA(ctx, theArg.tbaa);
+                ai.decorateInst(ctx.builder.CreateStore(emit_unbox(ctx, to, theArg, f.typ), argaddr));
+                theArg = mark_julia_slot(argaddr, f.typ, NULL, ctx.tbaa().tbaa_stack);
+            }
+            Value *specptr_addr = ctx.builder.CreateInBoundsGEP(
+                    getInt8Ty(ctx.builder.getContext()), argaddr,
+                    ConstantInt::get(ctx.types().T_size, offsetof(jl_opaque_closure_t, specptr)));
+
+            jl_cgval_t closure_specptr = typed_load(ctx, specptr_addr, NULL, (jl_value_t*)jl_long_type,
+                theArg.tbaa, nullptr, false, AtomicOrdering::NotAtomic, false, ctx.types().alignof_ptr.value());
+
+            Value *specptr = emit_unbox(ctx, ctx.types().T_size, closure_specptr, (jl_value_t*)jl_long_type);
```

Comment:
Probably might as well right this with the LLVM API directly, since you are bypassing the Julia API already anyways to compute the field offset and this is adding a lot of boilerplate to recover that.
```
    Value *specptr = ctx.builder.CreateAlignedLoad(ctx.types().T_size, emit_bitcast(ctx, specptr_addr, ctx.types().T_psize), ctx.types().alignof_ptr.value());
```